File: src/main/csharp/de/kimminich/agile/exercises/lecture5/SmartAIStrategyTest.cs
namespace de.kimminich.agile.exercises.lecture5
{

	using NUnit.Framework.Constraints;
	using Card = de.kimminich.agile.exercises.lecture2.Card;
	using ICardChooserStrategy = de.kimminich.agile.exercises.lecture3.ICardChooserStrategy;
	using Player = de.kimminich.agile.exercises.lecture2.Player;
	using PlayerHelper = de.kimminich.agile.exercises.lecture5.helper.PlayerHelper;

	[TestFixture]
	public class SmartAIStrategyTest : AssertionHelper
	{

		internal ICardChooserStrategy strategy = (player, opponent) => null; // TODO Instantiate implementation to be tested instead

		[Test]
		public void shouldNotKeepAnyZeroCardsInHand()
		{
			Player player = new Player();
			PlayerHelper.putCardsInHandOf(player, 0, 0, 1, 2, 3);

			Player opponent = new Player();

			Card cardToPlay;
			while ((cardToPlay = strategy.nextCardToPlay(player, opponent)) != null)
			{
				PlayerHelper.simulatePlayingCard(player, cardToPlay);
			}

			Assert.AreEqual(0, countOccurencesOfCardInHand(player, 0));
		}

		[Test]
		public void shouldPreferToPlayMultipleLowCardsOverSingleHighCardToPreventOverload()
		{
			Player player = new Player();
			player.Mana = 7;
			PlayerHelper.putCardsInHandOf(player, 1, 2, 4, 7, 6);

			Player opponent = new Player();

			Card cardToPlay;
			while ((cardToPlay = strategy.nextCardToPlay(player, opponent)) != null)
			{
				PlayerHelper.simulatePlayingCard(player, cardToPlay);
			}

			Assert.AreEqual(1, countOccurencesOfCardInHand(player, 7));
			Assert.AreEqual(1, countOccurencesOfCardInHand(player, 6));
			Assert.AreEqual(0, countTotalOccurencesOfCardsInHand(player, 1, 2, 4));
		}

	}

}

File: src/main/java/de/kimminich/agile/exercises/lecture5/tcg/SmartAIStrategyTest.java
package de.kimminich.agile.exercises.lecture5.tcg;

import de.kimminich.agile.exercises.lecture2.tcg.Player;
import de.kimminich.agile.exercises.lecture3.tcg.ICardChooserStrategy;
import org.junit.Test;

import static de.kimminich.agile.exercises.lecture5.tcg.helper.PlayerHelper.*;
import static org.junit.Assert.assertEquals;

public class SmartAIStrategyTest {

    ICardChooserStrategy strategy = (player, opponent) -> null; // TODO Instantiate implementation to be tested instead

    @Test
    public void shouldNotKeepAnyZeroCardsInHand() {
        Player player = new Player();
        putCardsInHandOf(player, 0, 0, 1, 2, 3);

        Player opponent = new Player();

        Integer cardToPlay;
        while ((cardToPlay = strategy.nextCardToPlay(player, opponent)) != null) {
            simulatePlayingCard(player, cardToPlay);
        }

        assertEquals(0, countOccurencesOfCardInHand(player, 0));
    }

    @Test
    public void shouldPreferToPlayMultipleLowCardsOverSingleHighCardToPreventOverload() {
        Player player = new Player();
        player.setMana(7);
        putCardsInHandOf(player, 1, 2, 4, 7, 6);

        Player opponent = new Player();

        Integer cardToPlay;
        while ((cardToPlay = strategy.nextCardToPlay(player, opponent)) != null) {
            simulatePlayingCard(player, cardToPlay);
        }

        assertEquals(1, countOccurencesOfCardInHand(player, 7));
        assertEquals(1, countOccurencesOfCardInHand(player, 6));
        assertEquals(0, countTotalOccurencesOfCardsInHand(player, 1, 2, 4));
    }

}
